# Projet du cours "Kernel methods" du master MVA

Il s'agit du code associé au kernel Polynomial, moins détaillé que le fichier RBF qui présente les résultats principaux, mais contient tout de même plus de précisions que le kernel Linear.

## Chargement des modules

In [1]:
%pylab inline
import pandas
import cvxopt
import time
import numpy as np
import numpy.linalg as la
import scipy
from scipy.stats import mode
import matplotlib.pyplot as plt 

Populating the interactive namespace from numpy and matplotlib


# Importation des données

On importe les données du concours. On peut aussi insérer à la place de Xtr la base obtenue en appliquant les ondelettes pour voir les résultats obtenus avec les ondelettes.

In [2]:
df_train=pandas.read_csv('/Users/badr-eddinecherief-abdellatif/Documents/ENSAE/3A/MVA/Kernel methods/Projet ENS/Bases/Xtr.csv',header=None,sep=',')
df_train=df_train.drop(3072,1)
df_test=pandas.read_csv('/Users/badr-eddinecherief-abdellatif/Documents/ENSAE/3A/MVA/Kernel methods/Projet ENS/Bases/Xte.csv',header=None,sep=',')
df_test=df_test.drop(3072,1)
y_train=pandas.read_csv('/Users/badr-eddinecherief-abdellatif/Documents/ENSAE/3A/MVA/Kernel methods/Projet ENS/Bases/Ytr.csv',sep=';')
y_train=pandas.DataFrame(y_train['Prediction'])

# Répartition des labels
Nous travaillons sur un échantillon d'apprentissage dans le but de classifier des images. Vérifions que nous avons bien équirépartition des images dans l'échantillon d'apprentissage afin de pouvoir mener à bien notre modélisation

In [3]:
y_train["Prediction"].value_counts()

7    500
3    500
6    500
2    500
9    500
5    500
1    500
8    500
4    500
0    500
Name: Prediction, dtype: int64

# Implémentation des méthodes

Dans cette partie, nous implémenterons les différentes méthodes que nous testerons par la suite. Il s'agit des méthodes suivantes:
- SVM
- ACP
- HOG


## Support Vector Machine

In [4]:
class SVMTrainer(object):
    def __init__(self, c, degree, constant):
        self._c = c
        self._degree = degree
        self._constant = constant
        
    def train(self, X, y):
        lagrange_multipliers = self._compute_multipliers(X, y)
        return self._construct_predictor(X, y, lagrange_multipliers)

    def _gram_matrix(self, X):
        n_samples, n_features = X.shape
        K = ( [[self._constant]*n_samples]*n_samples + dot(X.values,transpose(X.values)) ) ** self._degree
        print("Taille de la matrice de Gram :",K.shape)
        return K

    def _construct_predictor(self, X, y, lagrange_multipliers):
        support_vector_indices=abs(lagrange_multipliers)>0
        support_multipliers = lagrange_multipliers[support_vector_indices] # alpha i positifs
        support_vectors = X[support_vector_indices] # vecteurs support
        support_vector_labels = y[support_vector_indices] # labels des vecteurs support
        print("Nombre de vecteurs supports :",len(support_vectors))
        pre=SVMPredictor(degree=self._degree,constant=self._constant,bias=0.0,weights=support_multipliers,
             support_vectors=support_vectors,support_vector_labels=support_vector_labels).predictest(support_vectors.values)[0]
        bias = np.mean(support_vector_labels[:,0]-pre)
        print("Biais:",bias)
        return SVMPredictor(degree=self._degree,constant=self._constant,bias=bias,weights=support_multipliers,support_vectors=support_vectors,
            support_vector_labels=support_vector_labels)

    def _compute_multipliers(self, X, y):
        n_samples, n_features = X.shape
        K=self._gram_matrix(X)
        P=cvxopt.matrix(K)
        q=cvxopt.matrix(-1*y[:,0])
        G_std=cvxopt.matrix(-1*np.diag(y[:,0]))
        h_std=cvxopt.matrix(np.zeros(n_samples))
        G_slack=cvxopt.matrix(np.diag(y[:,0]))
        h_slack=cvxopt.matrix(np.ones(n_samples) * self._c)
        G=cvxopt.matrix(np.vstack((G_std, G_slack)))
        h=cvxopt.matrix(np.vstack((h_std, h_slack)))
        A=cvxopt.matrix(np.ones(n_samples),(1,n_samples))
        b=cvxopt.matrix(0.0)
        solution = cvxopt.solvers.qp(P, q, G, h, A=None, b=None)
        return np.ravel(solution['x'])

In [5]:
class SVMPredictor(object):
    def __init__(self,degree,constant,bias,weights,support_vectors,support_vector_labels):
        self._degree = degree
        self._constant = constant
        self._bias = bias
        self._weights = weights
        self._support_vectors = support_vectors
        self._support_vector_labels = support_vector_labels

    def predictest(self,x):
        inter = dot(self._support_vectors,transpose(x))
        for p in range(inter.shape[0]) :
            for q in range(inter.shape[1]) :
                inter[p][q]=self._constant
        inter1 = ( inter + dot(self._support_vectors,transpose(x)) ) ** self._degree
        result = dot(transpose(self._weights),inter1)+self._bias
        return sign(result),result

## Adaptation du SVM au cas multi-classe
On va tout d'abord implémenter la méthode One Vs One, i.e. considérer 10x9/2=45 problèmes de classifications binaires en traitant des paires d'instances, puis implémenter la méthode One Vs Rest, i.e. entraîner un classifieur par classe.

### One Vs One

In [6]:
def SVMOvO(data_train,data_test,is_train,C,degree,constant):
    SVM=SVMTrainer(C,degree,constant)
    result=(20*ones(len(data_test))).reshape(len(data_test),1)
    for i in [0,1,2,3,4,5,6,7,8]:
        for j in range(i+1,10):
            print("Classe :",i,j)
            df_train_ij=data_train[asmatrix(is_train,dtype=ndarray)==(i,j)].copy()
            is_train_ij=is_train[asmatrix(is_train,dtype=ndarray)==(i,j)].copy()
            is_train_ij[is_train==i]=1
            is_train_ij[is_train==j]=-1
            s1=time.clock()
            SVMPred=SVM.train(df_train_ij,asfarray(is_train_ij))
            s2=time.clock()
            print("Temps pour l'entrainement :",s2-s1)
            s3=time.clock()
            values=SVMPred.predictest(data_test.values)[0]
            s4=time.clock()
            print("Temps pour la prédiction :",s4-s3)
            values[values==1]=i
            values[values==-1]=j
            values=values.reshape(len(data_test),1)
            result=concatenate((result,values),1)
            is_test_predict=mode(result,axis=1)
    return(is_test_predict)    

### One Vs Rest

In [7]:
def SVMOvR(data_train,data_test,is_train,C,degree,constant):
    SVM=SVMTrainer(C,degree,constant)
    result=(-20*ones(len(data_test))).reshape(len(data_test),1)
    for i in [0,1,2,3,4,5,6,7,8,9]:
        print("Classe:",i)
        is_train_i=is_train.copy()
        is_train_i[is_train==i]=1
        is_train_i[is_train!=i]=-1
        s1=time.clock()
        SVMPred=SVM.train(data_train,asfarray(is_train_i))
        s2=time.clock()
        print("Temps pour l'entrainement:",s2-s1)
        s3=time.clock()
        values=SVMPred.predictest(data_test.values)[1]
        s4=time.clock()
        print("Temps pour la prédiction:",s4-s3)
        values=values.reshape(len(data_test),1)
        result=concatenate((result,values),1)
    predict=argmax(result,axis=1)-1
    return(predict)  

# Analyse en Composantes Principales (non-kernel)
Il s'agit d'une ACP toute simple, où on détermine les valeurs propres et vecteurs propres correspondant à la matrice de variance de la base de données.

In [8]:
class ACP(object):
    def __init__(self, n_components):
        self.n_components = n_components
         
    def center(self,X):
        result = X - mean(X,axis=0)
        return result

    def vpropres(self,X):
        n_samples,n_var = X.shape
        X_centre = self.center(X)
        Var = (1/n_samples)*dot(X_centre.T,X_centre)
        de = diag(var(Var,axis=1))
        cool = dot(Var,transpose(de))
        vp = scipy.linalg.eig(cool)
        a = sorted(vp[0],reverse=True)
        stop = a[self.n_components-1]
        vect = vp[1][:,vp[0]>=stop]
        print("Variance expliquée :" , real(sum(vp[0][vp[0]>=stop]))/real(sum(vp[0])))
        return(vect)
    
    def fit_transform(self,X):
        X_centre = self.center(X)
        vect = self.vpropres(X_centre)
        X_transform = dot(X,vect)
        return(pandas.DataFrame(real(X_transform)))

# Tests

Afin de tester nos modèles sur nos données d'entraînement, on va découper notre base d'entraînement en train/test à hauteur de 60/40%, et effectuer les tests avec chaque méthode sans HOG. Nous avons constaté après tests que les résultats sont plus mauvais qu'avec le noyau RBF, et nous n'approfondirons donc pas ce noyau et nous nous centrerons donc sur les autres. Nous présentons les tests sans HOG et avec la méthode SVM One-Vs-One.

In [9]:
X_train_1 = df_train[0:3000]
X_test_1 = df_train[3000:5000]
Y_train_1 = y_train[0:3000]
Y_test_1 = y_train[3000:5000]
print(X_train_1.shape)
print(X_test_1.shape)
Y_train_1["Prediction"].value_counts()

(3000, 3072)
(2000, 3072)


3    311
1    308
0    306
6    305
8    304
4    298
5    294
7    292
9    291
2    291
Name: Prediction, dtype: int64

## Test SVM One Vs One

In [15]:
y_predict_OvO=SVMOvO(X_train_1,X_test_1,Y_train_1,6,1,1)

Classe : 0 1
Taille de la matrice de Gram : (614, 614)
     pcost       dcost       gap    pres   dres
 0:  5.7542e+02 -1.6433e+04  4e+04  7e-01  1e-14
 1:  6.9641e+02 -4.9476e+03  6e+03  3e-03  2e-14
 2:  5.9778e+01 -9.5944e+02  1e+03  4e-04  1e-14
 3: -9.7355e+01 -1.8730e+02  9e+01  2e-16  1e-14
 4: -1.1119e+02 -1.2578e+02  1e+01  2e-16  7e-15
 5: -1.1370e+02 -1.1564e+02  2e+00  2e-16  7e-15
 6: -1.1415e+02 -1.1427e+02  1e-01  2e-16  7e-15
 7: -1.1419e+02 -1.1420e+02  5e-03  2e-16  7e-15
 8: -1.1419e+02 -1.1419e+02  2e-04  2e-16  7e-15
 9: -1.1419e+02 -1.1419e+02  8e-06  2e-16  7e-15
Optimal solution found.
Nombre de vecteurs supports : 614
Biais: 0.0
Temps pour l'entrainement : 1.2016319999999894
Temps pour la prédiction : 1.2605220000000088
Classe : 0 2
Taille de la matrice de Gram : (597, 597)
     pcost       dcost       gap    pres   dres
 0:  4.9463e+02 -1.3800e+04  2e+04  4e-01  2e-14
 1:  2.7405e+02 -2.1610e+03  2e+03  2e-16  2e-14
 2: -1.2454e+02 -5.4281e+02  4e+02  2e-16  1

In [16]:
compt=Y_test_1.values-y_predict_OvO.mode
print("Pourcentage de bonnes prédictions :",sum(compt==0)/len(Y_test_1))

Pourcentage de bonnes prédictions : 0.174


## Test SVM One Vs One après ACP

### On applique une ACP et on conserve 30 composantes principales 

In [10]:
acp=ACP(30)
X_train_acp=acp.fit_transform(df_train)
print(X_train_acp.shape)

Variance expliquée : 0.220212319462
(5000, 30)


In [18]:
X_train_2 = X_train_acp[0:3000]
X_test_2 = X_train_acp[3000:5000]

In [19]:
y_predict_OvO_ACP=SVMOvO(X_train_2,X_test_2,Y_train_1,6,1,1)

Classe : 0 1
Taille de la matrice de Gram : (614, 614)
     pcost       dcost       gap    pres   dres
 0: -1.5944e+03 -2.4703e+04  5e+04  9e-01  3e-14
 1: -1.4926e+03 -9.3737e+03  8e+03  2e-16  4e-14
 2: -1.8938e+03 -3.1565e+03  1e+03  2e-16  4e-14
 3: -2.2785e+03 -2.7557e+03  5e+02  2e-16  4e-14
 4: -2.4181e+03 -2.5834e+03  2e+02  2e-16  5e-14
 5: -2.4612e+03 -2.5309e+03  7e+01  2e-16  5e-14
 6: -2.4827e+03 -2.5040e+03  2e+01  2e-16  6e-14
 7: -2.4879e+03 -2.4976e+03  1e+01  2e-16  5e-14
 8: -2.4908e+03 -2.4942e+03  3e+00  2e-16  6e-14
 9: -2.4920e+03 -2.4927e+03  7e-01  2e-16  6e-14
10: -2.4922e+03 -2.4925e+03  3e-01  2e-16  5e-14
11: -2.4923e+03 -2.4923e+03  2e-02  2e-16  6e-14
12: -2.4923e+03 -2.4923e+03  4e-04  2e-16  6e-14
Optimal solution found.
Nombre de vecteurs supports : 614
Biais: -0.159609120521
Temps pour l'entrainement : 1.3086759999999913
Temps pour la prédiction : 1.0719740000000115
Classe : 0 2
Taille de la matrice de Gram : (597, 597)
     pcost       dcost       ga

In [20]:
compt=Y_test_1.values-y_predict_OvO_ACP.mode
print("Pourcentage de bonnes prédictions:",sum(compt==0)/len(Y_test_1))

Pourcentage de bonnes prédictions: 0.165


## Optimisation des paramètres

In [21]:
y_predict_OvO = [SVMOvO(X_train_1,X_test_1,Y_train_1,1,1,1)]*50
for degree in range(5) :
    for constant in range(10) :
        y_predict_OvO[degree*10+constant] = SVMOvO(X_train_1,X_test_1,Y_train_1,1,degree,constant)

Classe : 0 1
Taille de la matrice de Gram : (614, 614)
     pcost       dcost       gap    pres   dres
 0:  5.7542e+02 -1.6433e+04  4e+04  7e-01  1e-14
 1:  6.9641e+02 -4.9476e+03  6e+03  3e-03  2e-14
 2:  5.9778e+01 -9.5944e+02  1e+03  4e-04  1e-14
 3: -9.7355e+01 -1.8730e+02  9e+01  2e-16  1e-14
 4: -1.1119e+02 -1.2578e+02  1e+01  2e-16  7e-15
 5: -1.1370e+02 -1.1564e+02  2e+00  2e-16  7e-15
 6: -1.1415e+02 -1.1427e+02  1e-01  2e-16  7e-15
 7: -1.1419e+02 -1.1420e+02  5e-03  2e-16  7e-15
 8: -1.1419e+02 -1.1419e+02  2e-04  2e-16  7e-15
 9: -1.1419e+02 -1.1419e+02  8e-06  2e-16  7e-15
Optimal solution found.
Nombre de vecteurs supports : 614
Biais: 0.0
Temps pour l'entrainement : 1.2616179999999986
Temps pour la prédiction : 1.7326160000000073
Classe : 0 2
Taille de la matrice de Gram : (597, 597)
     pcost       dcost       gap    pres   dres
 0:  4.9463e+02 -1.3800e+04  2e+04  4e-01  2e-14
 1:  2.7405e+02 -2.1610e+03  2e+03  2e-16  2e-14
 2: -1.2454e+02 -5.4281e+02  4e+02  2e-16  1

In [22]:
for constant in range(10) :
    for degree in range(5) :
        compt=Y_test_1.values-y_predict_OvO[degree*10+constant].mode
        print( (degree,constant) , sum(compt==0)/len(Y_test_1))

(0, 0) 0.0945
(1, 0) 0.1715
(2, 0) 0.2595
(3, 0) 0.218
(4, 0) 0.206
(0, 1) 0.0945
(1, 1) 0.174
(2, 1) 0.2145
(3, 1) 0.1835
(4, 1) 0.133
(0, 2) 0.0945
(1, 2) 0.1745
(2, 2) 0.208
(3, 2) 0.194
(4, 2) 0.1675
(0, 3) 0.0945
(1, 3) 0.1745
(2, 3) 0.1955
(3, 3) 0.191
(4, 3) 0.1835
(0, 4) 0.0945
(1, 4) 0.1745
(2, 4) 0.191
(3, 4) 0.199
(4, 4) 0.186
(0, 5) 0.0945
(1, 5) 0.1745
(2, 5) 0.194
(3, 5) 0.2015
(4, 5) 0.1885
(0, 6) 0.0945
(1, 6) 0.1745
(2, 6) 0.19
(3, 6) 0.1955
(4, 6) 0.1935
(0, 7) 0.0945
(1, 7) 0.1745
(2, 7) 0.1895
(3, 7) 0.194
(4, 7) 0.1975
(0, 8) 0.0945
(1, 8) 0.1745
(2, 8) 0.187
(3, 8) 0.1915
(4, 8) 0.1985
(0, 9) 0.0945
(1, 9) 0.1745
(2, 9) 0.186
(3, 9) 0.1885
(4, 9) 0.196


In [23]:
## Paramètres optimaux
C = 1
degree = 2
constant = 0

A priori, les paramètres précédents donnent des résultats très satisfaisants (25.95%) qui pourrait nous tenter de conserver ce noyau pour le modèle final. Toutefois, nous avons constaté que le noyau RBF donnait de meilleurs prédictions après application de la méthode HOG, et c'est la raison pour laquelle nous avons finalement appliqué ce noyau au lieu du noyau Polynomial.

# Comparaison avec le package Scikit-Learn

In [24]:
from sklearn.svm import SVC
clf = SVC(kernel="rbf",C=6,gamma=1/(2*2**2),degree=2)
clf.fit(X_train_1,Y_train_1)
compt = Y_test_1.values[:,0] - clf.predict(X_test_1)
print("Pourcentage de bonnes prédictions :",sum(compt==0)/len(Y_test_1))

//anaconda/lib/python3.5/site-packages/sklearn/svm/base.py:514: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y_ = column_or_1d(y, warn=True)


Pourcentage de bonnes prédictions : 0.2115


# Les résultats sont moins bons que pour le kernel RBF. 

# Nous poursuivrons l'étude avec le noyau RBF.